src/librbd/Operations.cc

```diff
@@ -333,7 +346,48 @@ bool needs_invalidate(I& image_ctx, uint64_t object_no,
 
 template <typename I>
 Operations<I>::Operations(I &image_ctx)
-  : m_image_ctx(image_ctx), m_async_request_seq(0) {
+  : m_image_ctx(image_ctx), m_async_request_seq(0),
+    m_queue_lock(ceph::make_mutex(
+                   util::unique_lock_name("librbd::Operations::m_queue_lock",
+                                          this))) {
+}
+
+template <typename I>
+void Operations<I>::queue(const std::string &name, Context *ctx) {
+  ctx = util::create_async_context_callback(
+      m_image_ctx, new LambdaContext(
+        [this, name, ctx](int r) {
+          if (r == 0) {
+            std::shared_lock owner_locker{m_image_ctx.owner_lock};
+            std::shared_lock image_locker{m_image_ctx.image_lock};
+
+            if (m_image_ctx.exclusive_lock != nullptr &&
+                !m_image_ctx.exclusive_lock->is_lock_owner()) {
+              r = -ERESTART;
+            }
+          }
+
+          ctx->complete(r);
+
+          std::unique_lock locker{m_queue_lock};
+          auto &queue = m_queued_ops[name];
+          if (queue.empty()) {
+            m_in_flight_ops.erase(name);
+            return;
+          }
+
+          auto queued_ctx = queue.front();
+          queue.pop_front();
+          queued_ctx->complete(r);
```

Comment:
Nit: should you drop the lock before invoking?

Reply:
The context is wrapped with `util::create_async_context_callback` so I think it is not necessary to drop the lock.